html export: export project_map.js, add stable_uri_forwarder.js

thseiler opened the pull request:
This PR addresses #2100 and adds:
- export of the project structure as /assets/project_map.js
- stable_uri_forwarder.js
- integration and end2end tests

How to use:
1) start strictdoc server or do a strictdoc export of the project
2) open project_index URL in browser and add UID or MID as anchor, e.g. http://localhost:5111/#SOME_UID
3) browser jumps to the given node within the project

Not yet there:
- UI to copy the requirement's stable link to the clipboard...

thseiler committed:
html export: #2100 export project_map.js, add stable_uri_forwarder.js

stanislaw commented:
@thseiler this is super cool 🎉 🎉 🎉 ! I have three initial comments:

- Please rebase your branch and push again. I was doing some refactoring and we ran into a race condition – I had fixed something after my change but you pushed before my fix was there.
- FYI: While reviewing your integration test #4, I noticed a small unrelated regression which I am fixing here: https://github.com/strictdoc-project/strictdoc/pull/2145.
- I don't see that this forwarder is working in the static HTML. Could you confirm that it should work or you only tested with the web server? I didn't spend enough time to understand this myself yet.

thseiler commented:
Thanks! after rebasing, the CI tests now passed :-) Nice catch for #2145. 
I suspect Safari might set scriptUrl.origin to 'null' for scripts loaded from the local filesystem (file:///). So I changed the forwarder script to use URL objects for the path wrangling. This should make the script more portable. I tested with Chrome, Edge, and Firefox on Windows, and the forwarding works for me on both static HTML export and the web server. Could you give it another try with the latest push? Thanks!

stanislaw commented:
> I suspect Safari might set scriptUrl.origin to 'null' for scripts loaded from the local filesystem (file:///). So I changed the forwarder script to use URL objects for the path wrangling. This should make the script more portable. I tested with Chrome, Edge, and Firefox on Windows, and the forwarding works for me on both static HTML export and the web server. Could you give it another try with the latest push? Thanks!

I tried the latest branch with Safari and the static HTML works in both Safari and Firefox.

However, I don't see it working for both `http://localhost:5111/index.html#SDOC_UG_CONTACT` and `http://localhost:5111#SDOC_UG_CONTACT`. I don't see any JS errors, it just does nothing when I open such a link.

At first I thought that it could be related to the fact that StrictDoc's own documentation uses --host 127.0.0.1 but changing the host to localhost does not fix it.

You could try running `invoke server` on StrictDoc's repo and let me know if I am doing something wrong.

```
@task(aliases=["s"])
def server(context, input_path=".", config=None):
    assert os.path.isdir(input_path), input_path
    if config is not None:
        assert os.path.isfile(config), config
    config_argument = f"--config {config}" if config is not None else ""
    run_invoke_with_tox(
        context,
        ToxEnvironment.DEVELOPMENT,
        f"""
            python -m strictdoc.cli.main
                server {input_path} {config_argument}
                    --host localhost
                    --output-path ./output/cache/server
                    --reload
        """,
    )
```

Also, as per my comment [here](https://github.com/strictdoc-project/strictdoc/issues/2100#issuecomment-2707296584):

> A small addition to your considerations: I understand that this JS redirector makes more sense with static HTML. With the web UI, we could simply implement a route that will do the same thing much easier on the server side. Would you agree that the web UI should have a server route, otherwise the mostly server-side logic of the web UI will somewhat leak to JS.

I like a lot this PR for static HTML but for server, we could disable this static project map and simply do a FastAPI route that would handle both UID and MID and redirect with 302 to a correct document/UID path.

And regarding this:

> UI to copy the requirement's stable link to the clipboard...

Let @mettta and I think about the UI solution to this, and we will handle it shortly. I have opened a separate ticket for this for @mettta: https://github.com/strictdoc-project/strictdoc/issues/2148.

thseiler commented:
I am happy to hear that the static HTML forwarding also worked for Safari on Mac. The issue with SDOC_UG_CONTACT was an error in the project_map.js export. (In an attempt to keep the map's file size small, I had skipped over files with no requirements. This is of course wrong as these might still have sections with UIDs, as is the case with the user guide and SDOC_UG_CONTACT...)

I had exposed the project_map.js in the server for now to allow for end2end testing. But I agree that this was not clean. The server case needs a dynamic forwarder, as a node might be moved around, or could have its UID updated.

So:
- I fixed the project_map export to no longer skip over files with no requirements. 
- In the process, I learned that [DOCUMENT] can have MID and UIDs, so I export these now too.
- I added a new class "SDocTestHTMLExporter", to allow end2end tests against the exported HTML instead of the server.
- With these changes in place, I can end2end test the stable uri forwarding with the html export, so I dont could remove them from the web server case... 
- I couldn't start yet with the FastApi path (302 redirect), but hope to work on this in the evening.

Pushing this nevertheless, to see if the CI environement is happy with the new SDocTestHTMLExporter...

stanislaw commented on strictdoc/export/html/templates/screens/project_index/project_map_file.jinja:
```diff
@@ -0,0 +1,8 @@
+{%- set document = view_object.traceability_index.document_tree.get_document_by_path(file.get_full_path()) %}
+{%- set is_included = document.document_is_included() %}
```
It looks like this variable is now unused.

stanislaw commented on strictdoc/export/html/_static/stable_uri_forwarder.js:
```diff
@@ -0,0 +1,55 @@
+
```
Here, please add one-two sentences about the purpose if this JS file. Nothing fancy, just a small description.

stanislaw commented on strictdoc/export/html/generators/project_map.py:
```diff
@@ -0,0 +1,23 @@
+# mypy: disable-error-code="no-untyped-def"
```
Is it possible that you remove this mypy ignore. It is a new file, so you probably just need to add just a few types here and there.

stanislaw commented on strictdoc/export/html/html_generator.py:
```diff
@@ -487,6 +493,25 @@ def export_project_tree_screen(
+        *,
+        traceability_index: TraceabilityIndex,
+    ):
+        _assets_dir = os.path.join(
```
You probably don't need `_` here. Let's keep it at just `assets_dir`. Otherwise, it looks too unusual.

stanislaw commented on tests/end2end/exporter.py:
```diff
@@ -0,0 +1,97 @@
+            )
+        args.extend([self.path_to_tdoc_folder])
+
+        self.log_file_out = open(  # pylint: disable=consider-using-with
```
Since this is not using `with`, we should either use `exit_stack` like does it in `tests/end2end/server.py`, or use explicit close() calls in the close() method.

Otherwise, we will probably be leaking the file descriptors. I don't know how Python handles this but with sockets those would definitely not be closed after the process terminates.

thseiler replied:
D'oh! good catch!

stanislaw commented on tests/end2end/exporter.py:
```diff
@@ -0,0 +1,97 @@
+        )
+        self.path_to_err_log = os.path.join(
+            path_to_tmp_dir, "strictdoc_export.err.log"
+        )
```
Here somewhere `self.process = None` is missing and also let's add a type annotation to all variables.

stanislaw commented on tests/end2end/exporter.py:
```diff
@@ -0,0 +1,97 @@
+
+        sleep(test_environment.warm_up_interval_seconds)
+        print(  # noqa: T201
+            f"TestSDocHTMLExporter: "
```
Here you probably meant `SDocTestHTMLExporter`.

thseiler replied:
D'oh! Yes, my fingers were again typing faster than my brain...

thseiler commented:
Thanks a lot for the review, and thanks for the kind words. I have addressed the review findings, and surprisingly, the CI Environment is also happy with the SDocTestHTMLExporter. 

I'll rebase and then add the link export + 302 forwarding in the evening

stanislaw commented:
@thseiler please give us a heads up when it is ready for the next round of review. Is it already now?

stanislaw commented:
> @thseiler please give us a heads up when it is ready for the next round of review. Is it already now?

We have merged https://github.com/strictdoc-project/strictdoc/pull/2158 where we migrated some Stimulus JS-based controllers to static HTML. Previously it was not possible because we used a ES6 modules version of Stimulus. Now Stimulus is a so called UMD and we can definitely have it working with static HTML, not just web server.

The outcome of this is that the file where you added changes `strictdoc/export/html/_static/copy_stable_link_observer.js` was removed. What you need to do is to add these changes to the `copy_stable_link_button_controller.js`. As you already learned the code there is very similar and now it is the same file that works on both server and static HTML. The whole point of this change was to remove to JS files that do the same thing for static and server and have just one.

I could do this myself tomorrow by pushing to your branch (there are public holidays) or you could do it yourself.

stanislaw commented:
@thseiler I have integrated your work as-is with only minor changes by myself, please see here: https://github.com/strictdoc-project/strictdoc/pull/2159. I initially was still testing things and didn't want to rebase/force-push to this PR just in case. So I went with a separate PR.

This is fully integrated now and I will close it.
